src/core/qgsvectorlayerutils.cpp

```diff
@@ -418,7 +418,7 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
 
     // 4. passed attribute value
     // note - deliberately not using else if!
-    if ( !v.isValid() && attributes.contains( idx ) )
+    if ( attributes.contains( idx ) && !layer->primaryKeyAttributes().contains( idx ) )
```

Comment:
One silly question: does QGIS api treats primaryKeys as unique values? There is one exception to unique values, which is composed primary keys, because in this case the set is unique, not only the value. 

So, I agree with you that this check must be first, but I still feel that the primary key check must also be done, do you agree @m-kuhn ?

Reply:
For composed primary keys the logic is more complicated anyway, in these cases it's often that one part stays while the other one changes (e.g. "province" stays and "street_id" changes), but that's hard to know and needs custom logic / more configuration than possible right now. Let's ignore this therefore.

In simple primary key scenarios, it's the provider's responsibility to report PKs as unique. If it does not, it's something to fix on another level than here.

Reply:
Ok, I get it.

Can I put this over the first if?
```
if ( attributes.contains( idx ) 
         && !layer->primaryKeyAttributes().contains( idx )
         && !( fields.at( idx ).constraints().constraints() & QgsFieldConstraints::ConstraintUnique ) )
    {
      v = attributes.value( idx );
    }
```

And then delete that if that I changed in the pull request?